src/tools/compiletest/src/runtest.rs

```diff
@@ -1845,19 +1842,36 @@ impl<'test> TestCx<'test> {
             None => {}
         }
 
+        // Musl toolchain is build on linux-gnu host
+        // but with proper setup it can behave almost* like native linux-musl.
+        // One difference is "cc" which will link to glibc; force musl cc.
         if self.props.force_host {
             self.maybe_add_external_args(&mut rustc,
                                          self.split_maybe_args(&self.config.host_rustcflags));
+            if self.config.target.contains("musl") {
+                if let Some(ref linker) = self.config.linker {
+                    rustc.arg(format!("-Clinker={}", linker));
+                }
+            }
         } else {
             self.maybe_add_external_args(&mut rustc,
                                          self.split_maybe_args(&self.config.target_rustcflags));
             if !is_rustdoc {
                 if let Some(ref linker) = self.config.linker {
                     rustc.arg(format!("-Clinker={}", linker));
                 }
+            } else if self.config.target.contains("musl") {
+                if let Some(ref linker) = self.config.linker {
+                    rustc.arg(format!("--linker={}", linker));
+                }
             }
         }
 
+        // Use dynamic musl for tests because static doesn't allow creating dylibs
+        if self.config.target.contains("musl") {
+            rustc.arg("-Ctarget-feature=-crt-static");
```

Comment:
Does this change mean that we're not actually testing the static linkage of musl on the image any more? If so I think that's definitely something we want to keep testing on CI as it's one of the flagship use cases for musl

Reply:
Couldn't agree more but static musl [cannot build `dylibs` and `proc_macro`](https://github.com/rust-lang/rust/blob/cb6fafbdf3b09da28b09fa7e87a24658d77dd02f/src/librustc_codegen_utils/link.rs#L158-L170).
I believe different toolchain name for dynamic and static targets would would be great for this case.

Reply:
Maybe we should run tests twice? Once as dynamic host toolchain and once as `linux-gnu` target.

Reply:
Hm ok, I'll write a follow-up comment about testing